### PR TITLE
Specx workflow and NCL receipt

### PR DESCRIPTION
**Pull request by jukent**
## PR Summary

Demonstrates how to use the scipy.signal module for spectral analysis and a 1-to-1 NCL receipt.

## Related Tickets & Documents
Closes https://github.com/NCAR/geocat-comp/issues/124

## PR Checklist
**General**
- [x] PR includes a summary of changes
- [x] Link relevant issues, make one if none exist

### CODE REVIEW

**Comment by jukent** [comment]
I might be missing a couple places where I have to list these notebooks in order for them to visible. Do we have documentation on that yet?

**Comment by anissa111** [comment]
> I might be missing a couple places where I have to list these notebooks in order for them to visible. Do we have documentation on that yet?

Ahh, yeah. So ultimately I want to automate a lot of that, but for now, some of it is in the contributors guide.

File: applications/data_analysis/data_analysis.rst
Comment (cyschneck) [inline applications/data_analysis/data_analysis.rst:11]: Missing the `ipynb` extension, which I think is why it is not being included on the sidebar

File: ncl/ncl_index/ncl_index.rst
Comment (cyschneck) [inline ncl/ncl_index/ncl_index.rst:17]: Is this meant to be here?

File: ncl/ncl_index/ncl-index-table.csv
Comment (cyschneck) [inline ncl/ncl_index/ncl-index-table.csv:4]: Currently this link is broken as a `404`, it seems like the site can't find this directory

File: ncl/ncl_index/ncl_index.rst
Comment (cyschneck) [reply ncl/ncl_index/ncl_index.rst:17]: For `NCL_spectral_analysis.ipynb` to show up on the sidebar for `NCL Applications` is can be added to be `ncl_entries.rst` file

**Comment by cyschneck** [comment]
There are two main notebooks here: one that lives under data_analysis (`data_analysis/spectral_analysis.ipynb`) and the receipt and notebook that live as NCL files (`ncl/NCL_spectral_analysis.ipynb` and `ncl/receipts/NCL_receipt_spectral_alaysis`)

So spectral analysis should show up on the sidebar in two places, @anissa111 correct? One under `Applications > Data Analysis` and the other under `NCL Applications`

File: ncl/ncl_index/ncl-index-table.csv
Comment (cyschneck) [reply ncl/ncl_index/ncl-index-table.csv:4]: `NCL_spectral_analysis.ipynb` is currently living in `ncl` (that is where it is being [generated](https://ncar.github.io/geocat-applications/_preview/39/ncl/NCL_spectral_analysis.html)), should it be within `ncl/receipts`?

Comment (jukent) [reply ncl/ncl_index/ncl-index-table.csv:4]: No sorry, that was from before I had a separate NCL and NCL-receipts notebook.

**Comment by jukent** [comment]
Still unclear on the repetition of entries in NCL_index and NCL_applications

**Comment by anissa111** [comment]
> Still unclear on the repetition of entries in NCL_index and NCL_applications

Right, I think I'm also confused about what's going on here.

> There are two main notebooks here: one that lives under data_analysis (`data_analysis/spectral_analysis.ipynb`) and the receipt and notebook that live as NCL files (`ncl/NCL_spectral_analysis.ipynb` and `ncl/receipts/NCL_receipt_spectral_alaysis`)
> 
> So spectral analysis should show up on the sidebar in two places, @anissa111 correct? One under `Applications > Data Analysis` and the other under `NCL Applications`

@cyschneck, are you asking if the same physical notebook should be linked in two places (no, in my opinion) or are you asking if spectral analysis content should be in multiple different categories?

**Comment by cyschneck** [comment]
@anissa111, because there are two separate notebooks, they should both appear separately on the side bar? So I am asking `are you asking if spectral analysis content should be in multiple different categories?` Since this covers both a Python applications notebook and an NCL to Python notebook

File: ncl/ncl_index/ncl-index-table.csv
Comment (cyschneck) [inline ncl/ncl_index/ncl-index-table.csv:4]: Would it be useful to link to the specific functions [`specx_anal`](https://www.ncl.ucar.edu/Document/Functions/Built-in/specx_anal.shtml) and [`specxy_anal`](https://www.ncl.ucar.edu/Document/Functions/Built-in/specxy_anal.shtml) in the first column as well? In case people are searching for a specific function. @anissa111 if a notebook covers multiple functions, should each function get its own row (that links to the same notebook)? This is how I have trig functions set up

Comment (jukent) [reply ncl/ncl_index/ncl-index-table.csv]: Thanks addressed this in latest pushes.

File: ncl/ncl_index/ncl-index-table.csv
Comment (cyschneck) [inline ncl/ncl_index/ncl-index-table.csv:5]: There are extra marks on `scipy.signal.csd()` and `scipy.signal.coherence()`
```
scipy.signal.periodogram(), scipy.signal.convolve(), scipy.signal.csd()`, ``scipy.signal.coherence()
```

**Review by cyschneck — CHANGES_REQUESTED** [review]
Is `specx_anal` and `specxy_anal` duplicate notebooks? They can be one notebook that both link to. But I am not sure if that would work for the side bar under NCL Applications.

**Comment by jukent** [comment]
> Is `specx_anal` and `specxy_anal` duplicate notebooks? They can be one notebook that both link to

They are not duplicate. They use different data and perform different analysis.

File: ncl/ncl_index/ncl-index-table.csv
Comment (jukent) [reply ncl/ncl_index/ncl-index-table.csv]: Good eye @cyschneck ironically the extra "`" appeared because I was missing one.

**Comment by cyschneck** [comment]
Does the notebook need to be run to be able to view the plot or use `plt.show()`? The `Data Analysis > Spectral Analysis` have plots that are not visible currently

File: ncl/ncl_entries/specx_anal.ipynb
Comment (cyschneck) [inline ncl/ncl_entries/specx_anal.ipynb:1]: Would be useful to have `plt.show` as part of the notebooks so the graphs are visible

File: applications/data_analysis/spectral_analysis.ipynb
Comment (cyschneck) [inline applications/data_analysis/spectral_analysis.ipynb:1]: Would be useful to have `plt.show` as part of the notebooks so the graphs are visible here as well

File: ncl/ncl_entries/specxy_anal.ipynb
Comment (cyschneck) [inline ncl/ncl_entries/specxy_anal.ipynb:1]: Would be useful to have `plt.show` as part of the notebooks so the graphs are visible in this notebook as well

File: ncl/ncl_entries/specx_anal.ipynb
Comment (jukent) [reply ncl/ncl_entries/specx_anal.ipynb:1]: Thanks Cora, added this throughout and ran the cells.

**Comment by anissa111** [comment]
Getting back to this, I think it would be worth doing some higher-level updates before getting into some more fine-grained feedback.

I know it's been a minute since this has seen attention, so here are the things I noticed need updated from team discussions about since last time.

- Remove NCL references from python content (`applications/data_analysis/spectral_analysis.ipynb`)
- Take a pass through plotting/viz sections for non-essential lines (like `plt.tight_layout()`, probably `plt.show()`, possibly `figsize=(...)` if it doesn't effect the readability)
- Consider combining `ncl/ncl_entries/specx_anal.ipynb` and `ncl/ncl_entries/specxy_anal.ipynb` NCL entries and possibly reconfigure to include NCL content from the python content/receipt and to be more in line with the NCL template and other NCL entries.
- Make the receipt (`ncl/receipts/specx_anal.ipynb`) more consistent with receipt template and other reciepts (in the way of less narrative content, more testing, etc)

**Comment by anissa111** [comment]
Ah, gosh. I meant for that to show up as a commented review, my bad. I'm going to remove myself as a requested reviewer so that I can be re-requested/notified when it's is ready for another look.

**Comment by jukent** [comment]
figsize() and tightlayout() does affect readability on the subplots.

**Review by cyschneck — APPROVED** [review]

File: applications/data_analysis/spectral_analysis.ipynb
Comment (anissa111) [inline applications/data_analysis/spectral_analysis.ipynb:1]: <p>typo here "using using"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (anissa111) [inline applications/data_analysis/spectral_analysis.ipynb:1]: <p>Small thing, but I think there should be a comma after "currently".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (anissa111) [inline applications/data_analysis/spectral_analysis.ipynb:1]: <p>Small point "2-Dimensional" -&gt; "2-dimensional".</p><p> </p><p>Also, I genuinely don't know here, but should "Sea Level Pressure Anomalies" and "Periodogram" be capitalized?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (anissa111) [inline applications/data_analysis/spectral_analysis.ipynb:1]: <p>"too" -&gt; "two"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/39/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

**Review by anissa111 — COMMENTED** [review]
Also, I'm seeing some more of those mathjax things here:
image

File: ncl/ncl_entries/specx_specxy_anal.ipynb
Comment (kafitzgerald) [inline ncl/ncl_entries/specx_specxy_anal.ipynb]: Isn't this the tapering function for spectral analysis functions? https://github.com/NCAR/ncl/blob/edd46f5426286146313b31bd6826e3d3c360c0dc/ni/src/lib/nfpfort/specx_dp.f#L702

It is interesting that there seem to be a bunch of them in the codebase.

File: ncl/ncl_entries/specx_specxy_anal.ipynb
Comment (kafitzgerald) [inline ncl/ncl_entries/specx_specxy_anal.ipynb]:
```suggestion
    "For our 2-Dimensional spectral data we'll look at `SLP_DarwinTahitiAnom` which contains Darwin and Tahiti sea level pressure anomalies between 1935-1998.\n",
```

File: ncl/ncl_entries/specx_specxy_anal.ipynb
Comment (kafitzgerald) [inline ncl/ncl_entries/specx_specxy_anal.ipynb]: This seems to be broken.

File: applications/data_analysis/spectral_analysis.ipynb
Comment (kafitzgerald) [inline applications/data_analysis/spectral_analysis.ipynb]:
```suggestion
    "This document shows a concise workflow using the `scipy.signal` module. To learn more about the functions and keyword arguments available in this package, read SciPy's [signal processing documentation](https://docs.scipy.org/doc/scipy/reference/signal.html)."
```

File: applications/data_analysis/spectral_analysis.ipynb
Comment (kafitzgerald) [inline applications/data_analysis/spectral_analysis.ipynb:57]:
```suggestion
    "Then, we use [`xarray.open_dataset()`](http://xarray.pydata.org/en/stable/generated/xarray.open_dataset.html) function to read the data as an xarray dataset.\n",
```

File: applications/data_analysis/spectral_analysis.ipynb
Comment (kafitzgerald) [inline applications/data_analysis/spectral_analysis.ipynb:549]:
```suggestion
    "You can specify a `window` argument for smoothing your spectrum before spectral analysis. Here is a [list of supported SciPy windows](https://docs.scipy.org/doc/scipy/reference/signal.windows.html). The default is `boxcar` which is a unit window equivalent to no tapering.\n",
```

File: applications/data_analysis/spectral_analysis.ipynb
Comment (kafitzgerald) [inline applications/data_analysis/spectral_analysis.ipynb]: which example above?  should this be 1-D?

File: ncl/ncl_entries/specx_specxy_anal.ipynb
Comment (kafitzgerald) [inline ncl/ncl_entries/specx_specxy_anal.ipynb]: Is this really 2D or should the description be something more like cross-spectral analysis?

**Review by kafitzgerald — COMMENTED** [review]
Overall this is looking really good.  

Some more general comments/questions (more specific comments inline):
* It looks like the outputs are still stored stored on this though - I think the plan is to clear those now that there's infra to run the notebooks.
* It looks like `specxy_anal` is in the NCL Index, but not the receipts.  I assume that's an item for later, but wanted to double check.

File: ncl/ncl_entries/specx_specxy_anal.ipynb
Comment (jukent) [reply ncl/ncl_entries/specx_specxy_anal.ipynb]: Would I point to this function as NCL's `specx_dp.f.DTAPERX` or do I drop the f?

File: ncl/ncl_entries/ncl_entries.rst
Comment (jukent) [reply ncl/ncl_entries/ncl_entries.rst:21]: I'll be making your changes in a separate push. For some reason it says 4 hours ago, but was posted 30 minutes ago (30 minutes after @anissa111 's suggestions which are merged and sometimes overlap with yours).

File: ncl/ncl_entries/specx_specxy_anal.ipynb
Comment (jukent) [reply ncl/ncl_entries/specx_specxy_anal.ipynb]: Okay I will make this change tomorrow

File: ncl/ncl_entries/specx_specxy_anal.ipynb
Comment (kafitzgerald) [reply ncl/ncl_entries/specx_specxy_anal.ipynb]: I think I would refer to it as the DTAPERX subroutine and link to the line(s) in the code.

File: ncl/ncl_entries/specx_specxy_anal.ipynb
Comment (kafitzgerald) [reply ncl/ncl_entries/specx_specxy_anal.ipynb]: To be clear, I don't think this should hold up merging anything.  It's just a thought.

**Review by kafitzgerald — APPROVED** [review]

**Pull request merged**
